paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/ProjectionRowData.java

```diff
@@ -0,0 +1,250 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.flink;
+
+import org.apache.paimon.types.RowType;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.DecimalData;
+import org.apache.flink.table.data.MapData;
+import org.apache.flink.table.data.RawValueData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.data.StringData;
+import org.apache.flink.table.data.TimestampData;
+import org.apache.flink.types.RowKind;
+
+import javax.annotation.Nullable;
+
+import java.io.Serializable;
+
+/**
+ * A {@link RowData} that provides a mapping view of the original {@link RowData} according to
+ * projection information.
+ */
+public class ProjectionRowData implements RowData, Serializable {
+    private final RowType producedDataType;
+    private final int[][] projectedFields;
+    private final int[] lastProjectedFields;
+    private transient RowData row;
+
+    ProjectionRowData(RowType producedDataType, int[][] projectedFields) {
+        this.producedDataType = producedDataType;
+        this.projectedFields = projectedFields;
+        this.lastProjectedFields = new int[projectedFields.length];
+        for (int i = 0; i < projectedFields.length; i++) {
+            this.lastProjectedFields[i] = projectedFields[i][projectedFields[i].length - 1];
+        }
+    }
+
+    public ProjectionRowData replaceRow(RowData row) {
+        this.row = row;
+        return this;
+    }
+
+    public static @Nullable ProjectionRowData copy(@Nullable ProjectionRowData rowData) {
+        if (rowData == null) {
+            return null;
+        }
+        return new ProjectionRowData(rowData.producedDataType, rowData.projectedFields);
+    }
+
+    @Override
+    public int getArity() {
+        return projectedFields.length;
+    }
+
+    @Override
+    public RowKind getRowKind() {
+        return row.getRowKind();
+    }
+
+    @Override
+    public void setRowKind(RowKind rowKind) {
+        row.setRowKind(rowKind);
+    }
+
+    @Override
+    public boolean isNullAt(int i) {
+        RowData rowData = extractInternalRow(i);
+        if (rowData == null) {
+            return true;
+        }
+        return rowData.isNullAt(lastProjectedFields[i]);
+    }
+
+    private @Nullable RowData extractInternalRow(int i) {
+        int[] projectedField = projectedFields[i];
+        RowData rowData = this.row;
+        RowType dataType = producedDataType;
+        for (int j = 0; j < projectedField.length - 1; j++) {
```

Comment:
This is just for `isNullAt`? The cost is too high!

Reply:
I tried adding a different implementation to ProjectionRowData. Now it would make a read/write from the original RowData to avoid nested mapping during reads. Please see if this is proper.